Update to TF-M v1.4.0

rwalton-arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Bring the latest changes from TF-M v1.4.0 into mbed-os.

Update the Nuvoton python scripts to bump the hard coded TF-M version.

Update locations of ARM_MUSCA target support in targets.json.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Patater committed:
tfm: Add mbedtls_ecc_group_to_psa.h to crypto_extra.h

Include mbedtls_ecc_group_to_psa.h from crypto_extra.h so that clients
of PSA within Mbed OS do not need to behave differently depending on
which PSA implementation they are using.

This solution is not ideal as it makes it more difficult to update the
TF-M-provided psa/crypto_extra.h. We'll have to see what other options
we have for including additional headers based on the Mbed OS
configuration.

LDong-Arm committed:
psa: Include mbedtls_svc_key_id.h for TF-M

We have added definitions that are needed by Mbed TLS's PSK key exchange
but missing from TF-M's PSA to `mbedtls_svc_key_id.h`. To pick up those
definitions, TF-M's `psa/crypto_values.h' needs to include
`mbedtls_svc_key_id.h`.

rwalton-arm committed:
"Update directory platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST"

rwalton-arm committed:
"Update secure binaries for [('ARM_MUSCA_B1', 'ARMCLANG')]"

rwalton-arm committed:
"Update secure binaries for [('ARM_MUSCA_S1', 'ARMCLANG')]"

rwalton-arm committed:
"Update secure binaries for [('NU_M2354', 'GNUARM')]"

rwalton-arm committed:
tools: tfm: Update bin_utils

rwalton-arm committed:
tfm: Move tfm_ns_interface.c

tfm_ns_interface.c is intended to be overriden by clients to support
different targets. We copy this file from upstream into the mbed-os
platform library. We also have a specific "strong" overridden version
for the NU_M2354 target, which is located in its target library.
Previously the implementations in the platform library were decorated
with __attribute__(weak), and we provided a strong definition for the
NU_M2354 target. This worked fine because of weak linking, the linker
will pick the first "strong" definition and use that, avoiding any ODR
violations. However, upstream have removed __attribute__(weak) from the
function definitions, which caused multiply defined symbol errors when
trying to build the NU_M2354 target.

To work around the above issue, we remove the common definition in the
platform library; instead we copy the file to the Musca B1 and Musca S1
target libaries. This means the appropriate tfm_ns_interface.c is only
included in the build when compiling for the specific target which uses
it.

rwalton-arm committed:
targets: ARM_MUSCA: Update tfm_target_names

The directory structure upstream has changed. Now ARM_MUSCA board
support has been moved under an "arm" subdirectory. Update targets.json
with the new locations.

LDong-Arm reviewed (dismissed):
@rwalton-arm Thanks, it looks good.

In #14977 I forgot to mention we need to ensure Mbed TLS compatibility. The way to check this is build mbed-os-example-tls for those three targets. So far I can think of the following that need to be done:
* [ ] Cherry-pick [compatibility patches in imported PSA headers](https://github.com/ARMmbed/mbed-os/commits/mbed-os-6.14.0/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/include). There are four patches in the link, but only two of them got overridden by the import scripts (032fe4a and e5230c9) so those two need to be cherry picked. (We need those patches because TF-M maintains its own PSA headers rather than using the ones from Mbed TLS, and headers from the two repos are always inconsistent no matter what revisions we check out...)
* [ ] [mbedtls/psa_util.h](https://github.com/ARMmbed/mbed-os/blob/756830e77652a37c911b8489e928f9a4fcaaddef/connectivity/mbedtls/include/mbedtls/psa_util.h#L126) from Mbed TLS 2.25 (which we have in Mbed OS) contains support for older algorithms (e.g. MD2, MD4, etc.), which have been dropped from the latest TF-M and Mbed TLS (see [TF-M 1.4 commit](https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/commit/interface/include/psa/crypto_values.h?h=TF-Mv1.4.0&id=359167d3a2fe109adb9fae3a086864489fdc1005) and [Mbed TLS 3.0 PR](https://github.com/ARMmbed/mbedtls/pull/4588)). Those algorithms are _not_ enabled in Mbed OS (see connectivity/mbedtls/include/mbedtls/config.h) so we're okay. But some configs in mbed-os-example-tls do [have MD4 enabled](https://github.com/ARMmbed/mbed-os-example-tls/blob/749049a4f03ebc35ae497ec8c2e94bb51785054d/benchmark/mbedtls_config.h#L21) and cause a build error, shall we disable it in the example? @Patater 
* [ ] `PSA_ALG_AEAD_WITH_TAG_LENGTH` has been replaced by `PSA_ALG_AEAD_WITH_SHORTENED_TAG` in the latest PSA (see [commit](https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/commit/interface/include/psa/crypto_values.h?h=TF-Mv1.4.0&id=13b27cf6a74062d76a5fce384ff8c77e9e2de9d2)) but Mbed TLS 2.25 uses the old one. Adding back the old macro for compatibility fixes the problem.

I only tried to build the "benchmark" example in mbed-os-example-tls on Musca S1 and not anything else. Please make sure to try all of them on those targets & toolchains, and we might spot more things to fix!

rwalton-arm committed:
PSA: Reinstate deprecated macro

PSA_ALG_AEAD_WITH_TAG_LENGTH has been replaced with
PSA_ALG_AEAD_WITH_SHORTENED_TAG upstream. We could just update
psa_util.h to use the new macro, but we still have some targets that
only support older versions of PSA, so we reinstate the removed macro.

LDong-Arm commented on connectivity/mbedtls/include/mbedtls/psa_util.h:
```diff
@@ -86,9 +86,9 @@ static inline psa_algorithm_t mbedtls_psa_translate_cipher_mode(
             return( PSA_ALG_ECB_NO_PADDING );
         case MBEDTLS_MODE_GCM:
-            return( PSA_ALG_AEAD_WITH_TAG_LENGTH( PSA_ALG_GCM, taglen ) );
+            return( PSA_ALG_AEAD_WITH_SHORTENED_TAG( PSA_ALG_GCM, taglen ) );
```
I'd suggest keeping this unchanged, and adding the old `PSA_ALG_AEAD_WITH_TAG_LENGTH` back to platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM//TARGET_TFM_LATEST/include/psa/crypto_values.h

We still have a TF-M v1.0 target CYTFM_064B0S2_4343W which has old PSA in platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM//TARGET_TFM_V1_0/ without the new `PSA_ALG_AEAD_WITH_SHORTENED_TAG` macro. It uses the same Mbed TLS.

LDong-Arm replied:
We can try to build the TLS examples with CYTFM_064B0S2_4343W too, though everything will eventually be tested by the CI

rwalton-arm replied:
Slightly OT: It would be ideal if we ran the "build examples" CI on every change automatically (i.e in a github action) then we could find these issues immediately instead of relying on manual testing.

LDong-Arm replied:
There are many examples and targets (especially targets), and some issues are limited to only one target so the testing is quite extensive.

LDong-Arm reviewed (approved):
LGTM

LDong-Arm reviewed (dismissed):
One more remark about region_defs.h as noted in https://github.com/ARMmbed/mbed-os-tf-m-regression-tests/pull/127
It should be a fairly small change.

rwalton-arm committed:
TFM: Remove BL2 macro workaround for mbed-cli1/ARMClang

We worked around an issue with mbed-cli1 not recognising the BL2 macro
from targets.json by adding patched versions of region_defs.h and
flash_layout.h for ARM_MUSCA targets. In the patched headers we defined
the BL2 macro to ensure it can be picked up by the ARM scatter files
that include the headers.

The current solution is not robust, because it means that the
aforementioned headers easily become out of date. A workaround of
defining the macros in the scatter files which need them was suggested
in https://github.com/ARMmbed/mbed-os/issues/14762

This commit applies the suggested changes to the ARM_MUSCA scatter
files.

rwalton-arm committed:
TFM: MUSCA: Import partition maps from upstream

Import the latest partition headers from upstream. We no longer need to
patch the headers to define the BL2 macro as we now define it in the
scatter files for the MUSCA targets.

LDong-Arm reviewed (approved):
LGTM
